bpf: Split TCP and non-TCP conntrack tables

joestringer opened the pull request:
Based on ~~#5450~~.

We've observed recently that it's common for bursty non-TCP flows to impact the operation of the datapath with respect to how TCP connections are handled, where for instance frequent legitimate bursts of DNS traffic could crowd out the table, preventing new legitimate TCP connections from establishing.

By splitting the TCP flows from non-TCP flows in the connection tracking tables, there should be better isolation guarantees for how such traffic flows could impact each other.

This piggybacks on recent refactoring in #5480, #5488, #5450 to minimize the set of changes and allow the set of CT maps to be generically updated without hardcoding references to specific maps around the codebase.

The metrics that were added in #5296 are extended to also include L4 protocol (`tcp` / `any`) to account for the split in tables, because there isn't currently any co-ordination between the CT garbage collection runs so it's difficult to properly set histograms or gauges based on two individual events occurring.

At the moment, the existing map path is retained at the same place but newer datapath will only use this for TCP connections. This allows upgrade/downgrade to keep the TCP connections around without interrupting them as the map doesn't need to be recreated. Long-lived UDP connections would be disrupted during upgrade.

Related: #5048

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5501)
<!-- Reviewable:end -->

joestringer committed:
bpf: Split connection tracking for TCP and non-TCP

Split the connection tracking of TCP connections into a separate table
from non-TCP, to improve the isolation between the flows. This means
that if there is a bursty set of UDP traffic, it will not prevent new
TCP connections from establishing.

Related: #5048

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
bpf: Relax verifier in error cases

Relax the verifier a bit to ensure that when a split in TCP/ANY L4 proto
is introduced in the connection tracking tables, the datapath does not
exceed any complexity limits.

Cases:
* Protocols that can't be handled by the connection tracker.
* On egress, when LB fails leading to packet drop for IPv4 traffic.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
ctmap: Refactor mapType into separate file

This just tidies things up and makes it easier to navigate ctmap.go.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
metrics: Add L4 protocol to CT/GC metrics

An upcoming commit will extend these to also split by TCP/ANY L4, so add
the ANY part in in this commit to reduce the changes necessary in the
subsequent commit.

Signed-off-by: Joe Stringer <joe@covalent.io>

tgraf commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -45,48 +45,27 @@ var (
+	MapNameAny4Global = MapNameAny4 + "global"
 
 	MapNumEntriesLocal  = 64000
 	MapNumEntriesGlobal = 1000000
```
I suggest we do a 600K TCP and 400K UDP split going to get started and start running that to see how it goes. We'll likely have to make it configurable in the end but that change is a bit more invasive.

joestringer replied:
Seems reasonable. If we do this, we will force CT map recreation on upgrade so might as well name them better on the filesystem too. (The alternative would involve leaving the existing map's details the same and only use it for TCP connections, which should allow upgrade/downgrade to persist the connections without interruption).

tgraf reviewed (dismissed):
My main concern is around upgrade and downgrade. An upgrade would simply recreate the maps and thus flush all existing state. That may be acceptable. Downgrade to 1.2 would re-create the all-in-one map I suppose. Downgrade to 1.1 would only work to 1.1.4 I assume?

joestringer committed:
ctmap: Split CT entry sizes by protocol

Set the TCP CT table size to 512Ki, and CT table for other protocols to
256Ki.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
daemon: Allow configuration of CT max entries

Extend the ConfigMap and the daemon's arguments / environment variables
to allow configuration of the maximum number of CT entries.

Signed-off-by: Joe Stringer <joe@covalent.io>
